-1 
Tested at: 2c84bc5678f5cb14cf5b94525f97da9181409c76
 I found an error when building: 
 
 <pre>>> Compiling  /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_4_X_2015-07-29-2300/src/RecoParticleFlow/PFProducer/src/PFElectronAlgo.cc 
>> Compiling  /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_4_X_2015-07-29-2300/src/RecoParticleFlow/PFProducer/src/PFMuonAlgo.cc 
>> Compiling  /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_4_X_2015-07-29-2300/src/RecoParticleFlow/PFProducer/src/PFPhotonAlgo.cc 
>> Compiling  /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_4_X_2015-07-29-2300/src/RecoParticleFlow/PFProducer/src/PhotonSelectorAlgo.cc 
/tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_4_X_2015-07-29-2300/src/RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc: In member function 'float PFEGammaAlgo::EvaluateSingleLegMVA(const pfEGHelpers::HeavyObjectCache*, const PFBlockRef&, const reco::Vertex&, unsigned int)':
/tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_4_X_2015-07-29-2300/src/RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc:789:34: error: 'const class GBRForest' has no member named 'GetAdaBoostClassifier'
   mvaValue = hoc->gbrSingleLeg_->GetAdaBoostClassifier(vars);
                                  ^
/tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_4_X_2015-07-29-2300/src/RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc: In member function 'float PFEGammaAlgo::calculate_ele_mva(const pfEGHelpers::HeavyObjectCache*, const PFEGammaAlgo::ProtoEGObject&, reco::PFCandidateEGammaExtra&)':
/tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_4_X_2015-07-29-2300/src/RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc:2173:28: error: 'const class GBRForest' has no member named 'GetAdaBoostClassifier'
       return hoc->gbrEle_->GetAdaBoostClassifier(vars);
</pre> 
 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10474/6844/summary.html

--- a/RecoParticleFlow/PFProducer/interface/PFEGammaAlgo.h
+++ b/RecoParticleFlow/PFProducer/interface/PFEGammaAlgo.h
@@ -45,6 +45,8 @@
 #include <forward_list>
 #include <unordered_map>
 
+#include "RecoParticleFlow/PFProducer/interface/PFEGammaHeavyObjectCache.h"
+
 class PFSCEnergyCalibration;
 class PFEnergyCalibration;
 
@@ -129,7 +131,7 @@ class PFEGammaAlgo {
   //constructor
   PFEGammaAlgo(const PFEGConfigInfo&);
   //destructor
-  ~PFEGammaAlgo(){delete tmvaReaderEle_; delete tmvaReader_;   };
+  ~PFEGammaAlgo(){ };
 
   void setEEtoPSAssociation(const edm::Handle<EEtoPSAssociation>& eetops) {
     eetops_ = eetops;
@@ -142,14 +144,16 @@ class PFEGammaAlgo {
     cfg_.primaryVtx = & primary;
   }
 
-  void RunPFEG(const reco::PFBlockRef&  blockRef,
-	       std::vector< bool >& active
-	       );
-
+  void RunPFEG(const pfEGHelpers::HeavyObjectCache* hoc,
+               const reco::PFBlockRef&  blockRef,
+               std::vector< bool >& active
+               );
+               
   //check candidate validity
-  bool isEGValidCandidate(const reco::PFBlockRef&  blockRef,
-			  std::vector< bool >&  active){
-    RunPFEG(blockRef,active);
+  bool isEGValidCandidate(const pfEGHelpers::HeavyObjectCache* hoc,
+                          const reco::PFBlockRef&  blockRef,
+                          std::vector< bool >&  active){
+    RunPFEG(hoc,blockRef,active);
     return (egCandidate_.size()>0);
   };
   
@@ -204,7 +208,8 @@ private:
 
   // functions:
   // this runs the functions below
-  void buildAndRefineEGObjects(const reco::PFBlockRef& block);
+  void buildAndRefineEGObjects(const pfEGHelpers::HeavyObjectCache* hoc,
+                               const reco::PFBlockRef& block);
 
   // build proto eg object using all available unflagged resources in block.
   // this will be kind of like the old 'SetLinks' but with simplified and 
@@ -246,7 +251,8 @@ private:
   // refining steps doing the ECAL -> track piece
   // this is the factorization of the old PF photon algo stuff
   // which through arcane means I came to understand was conversion matching  
-  void linkRefinableObjectECALToSingleLegConv(ProtoEGObject&);
+  void linkRefinableObjectECALToSingleLegConv(const pfEGHelpers::HeavyObjectCache* hoc,
+                                              ProtoEGObject&);
 
   // wax off
 
@@ -261,14 +267,16 @@ private:
   
 
   // things for building the final candidate and refined SC collections    
-  void fillPFCandidates(const std::list<ProtoEGObject>&, 
+  void fillPFCandidates(const pfEGHelpers::HeavyObjectCache* hoc,
+                        const std::list<ProtoEGObject>&, 
 			reco::PFCandidateCollection&,
 			reco::PFCandidateEGammaExtraCollection&);
   reco::SuperCluster buildRefinedSuperCluster(const ProtoEGObject&);
   
   // helper functions for that
 
-  float calculate_ele_mva(const ProtoEGObject&,
+  float calculate_ele_mva(const pfEGHelpers::HeavyObjectCache* hoc,
+                          const ProtoEGObject&,
 			  reco::PFCandidateEGammaExtra&);
   void fill_extra_info(const ProtoEGObject&,
 		       reco::PFCandidateEGammaExtra&);
@@ -290,8 +298,6 @@ private:
   std::vector< std::pair <unsigned int, unsigned int> > convGsfTrack_;
 
   PFEGConfigInfo cfg_;
-  
-  TMVA::Reader    *tmvaReaderEle_;
 
   const char  *mvaWeightFile_;
 
@@ -327,7 +333,7 @@ private:
                                               */ 
   //FOR SINGLE LEG MVA:					      
   const reco::Vertex  *  primaryVertex_;
-  TMVA::Reader *tmvaReader_;
+  //TMVA::Reader *tmvaReader_;
   const GBRForest *ReaderLC_;
   const GBRForest *ReaderGC_;
   const GBRForest *ReaderRes_;
@@ -383,9 +389,10 @@ private:
 //   std::vector<reco::SuperCluser> sCluster_;
   reco::PFCandidateEGammaExtraCollection egExtra_;  
 
-  float EvaluateSingleLegMVA(const reco::PFBlockRef& blockref, 
-			    const reco::Vertex& primaryvtx, 
-			    unsigned int track_index);
+  float EvaluateSingleLegMVA(const pfEGHelpers::HeavyObjectCache* hoc,
+                             const reco::PFBlockRef& blockref, 
+                             const reco::Vertex& primaryvtx, 
+                             unsigned int track_index);
 };
 
 #endif

--- a/RecoParticleFlow/PFProducer/interface/PFEGammaHeavyObjectCache.h
+++ b/RecoParticleFlow/PFProducer/interface/PFEGammaHeavyObjectCache.h
@@ -1,0 +1,25 @@
+#ifndef __RecoParticleFlow_PFProducer_pfEGHelpersHeavyObjectCache_h__
+#define __RecoEgamma_GsfElectronAlgos_pfEGHelpersHeavyObjectCache_h__
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "CondFormats/EgammaObjects/interface/GBRForest.h"
+#include <memory>
+
+namespace pfEGHelpers {
+  class HeavyObjectCache {
+  public:
+    HeavyObjectCache(const edm::ParameterSet&);
+    std::unique_ptr<const GBRForest> gbrEle_;
+    std::unique_ptr<const GBRForest> gbrSingleLeg_;
+  private:
+    // for electron mva
+    float lnPt_gsf, Eta_gsf, dPtOverPt_gsf, DPtOverPt_gsf, chi2_gsf, nhit_kf;
+    float chi2_kf, EtotPinMode, EGsfPoutMode, EtotBremPinPoutMode, DEtaGsfEcalClust;
+    float SigmaEtaEta, HOverHE, lateBrem, firstBrem;
+    // for single leg mva
+    float nlost, nlayers;
+    float chi2, STIP, del_phi,HoverPt, EoverPt, track_pt;
+  };
+}
+
+#endif

--- a/RecoParticleFlow/PFProducer/plugins/PFEGammaProducer.cc
+++ b/RecoParticleFlow/PFProducer/plugins/PFEGammaProducer.cc
@@ -42,7 +42,8 @@ namespace {
   typedef std::list< reco::PFBlockRef >::iterator IBR;
 }
 
-PFEGammaProducer::PFEGammaProducer(const edm::ParameterSet& iConfig):
+PFEGammaProducer::PFEGammaProducer(const edm::ParameterSet& iConfig,
+                                   const pfEGHelpers::HeavyObjectCache*):
   primaryVertex_(reco::Vertex()),
   ebeeClustersCollection_("EBEEClusters"),
   esClustersCollection_("ESClusters") {
@@ -111,17 +112,11 @@ PFEGammaProducer::PFEGammaProducer(const edm::ParameterSet& iConfig):
 
   
   algo_config. mvaWeightFileEleID
-    = iConfig.getParameter<std::string>("pf_electronID_mvaWeightFile");
+    = iConfig.getParameter<edm::FileInPath>("pf_electronID_mvaWeightFile").fullPath();
 
   algo_config.applyCrackCorrections
     = iConfig.getParameter<bool>("pf_electronID_crackCorrection");
-  
-  std::string path_mvaWeightFileEleID;
-
-  algo_config.mvaWeightFileEleID = 
-    edm::FileInPath ( algo_config.mvaWeightFileEleID.c_str() ).fullPath();
-     
-
+    
   //PFPhoton Configuration
 
   std::string path_mvaWeightFileConvID;
@@ -132,10 +127,8 @@ PFEGammaProducer::PFEGammaProducer(const edm::ParameterSet& iConfig):
   std::string path_mvaWeightFileRes;
 
   algo_config.mvaweightfile =
-    iConfig.getParameter<std::string>("pf_convID_mvaWeightFile");
-  algo_config.mvaConvCut = iConfig.getParameter<double>("pf_conv_mvaCut");
-  algo_config.mvaweightfile = 
-    edm::FileInPath ( algo_config.mvaweightfile.c_str() ).fullPath();  
+    iConfig.getParameter<edm::FileInPath>("pf_convID_mvaWeightFile").fullPath();
+  algo_config.mvaConvCut = iConfig.getParameter<double>("pf_conv_mvaCut");  
   algo_config.sumPtTrackIsoForPhoton = 
     iConfig.getParameter<double>("sumPtTrackIsoForPhoton");
   algo_config.sumPtTrackIsoSlopeForPhoton = 
@@ -369,7 +362,7 @@ PFEGammaProducer::produce(edm::Event& iEvent,
     // keep track of the elements which are still active.
     std::vector<bool> active( elements.size(), true );      
     
-    pfeg_->RunPFEG(blockref,active);
+    pfeg_->RunPFEG(globalCache(),blockref,active);
 
     if( pfeg_->getCandidates().size() ) {
       LOGDRESSED("PFEGammaProducer")

--- a/RecoParticleFlow/PFProducer/plugins/PFEGammaProducer.h
+++ b/RecoParticleFlow/PFProducer/plugins/PFEGammaProducer.h
@@ -48,11 +48,19 @@ This producer makes use of PFAlgo, the particle flow algorithm.
 */
 
 
-class PFEGammaProducer : public edm::stream::EDProducer<> {
+class PFEGammaProducer : public edm::stream::EDProducer<edm::GlobalCache<pfEGHelpers::HeavyObjectCache> > {
  public:
-  explicit PFEGammaProducer(const edm::ParameterSet&);
+  explicit PFEGammaProducer(const edm::ParameterSet&, const pfEGHelpers::HeavyObjectCache* );
   ~PFEGammaProducer();
   
+  static std::unique_ptr<pfEGHelpers::HeavyObjectCache> 
+    initializeGlobalCache( const edm::ParameterSet& conf ) {
+       return std::unique_ptr<pfEGHelpers::HeavyObjectCache>(new pfEGHelpers::HeavyObjectCache(conf));
+   }
+  
+  static void globalEndJob(pfEGHelpers::HeavyObjectCache const* ) {
+  }
+
   virtual void produce(edm::Event&, const edm::EventSetup&) override;
   virtual void beginRun(const edm::Run &, const edm::EventSetup &) override;
 

--- a/RecoParticleFlow/PFProducer/python/particleFlowEGamma_cfi.py
+++ b/RecoParticleFlow/PFProducer/python/particleFlowEGamma_cfi.py
@@ -32,13 +32,13 @@ particleFlowEGamma = cms.EDProducer("PFEGammaProducer",
     # Use electron identification in PFAlgo
     usePFElectrons = cms.bool(True),
     pf_electron_output_col=cms.string('electrons'),
-    pf_electronID_mvaWeightFile = cms.string('RecoParticleFlow/PFProducer/data/MVAnalysis_BDT.weights_PfElectrons23Jan_IntToFloat.txt'),
+    pf_electronID_mvaWeightFile = cms.FileInPath('RecoParticleFlow/PFProducer/data/MVAnalysis_BDT.weights_PfElectrons23Jan_IntToFloat.txt'),
 
     # Use Photon identification in PFAlgo (for now this has NO impact, algo is swicthed off hard-coded
     usePFPhotons = cms.bool(True),
     usePhotonReg=cms.bool(False),
     useRegressionFromDB=cms.bool(True),                                 
-    pf_convID_mvaWeightFile = cms.string('RecoParticleFlow/PFProducer/data/MVAnalysis_BDT.weights_pfConversionAug0411.txt'),        
+    pf_convID_mvaWeightFile = cms.FileInPath('RecoParticleFlow/PFProducer/data/MVAnalysis_BDT.weights_pfConversionAug0411.txt'),        
     pf_conv_mvaCut=cms.double(0.0),                                 
     pf_locC_mvaWeightFile=cms.string('RecoParticleFlow/PFProducer/data/TMVARegression_BDTG_PFClusterLCorr_14Dec2011.root'),
     pf_GlobC_mvaWeightFile=cms.string('RecoParticleFlow/PFProducer/data/TMVARegression_BDTG_PFGlobalCorr_14Dec2011.root'),

--- a/RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc
+++ b/RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc
@@ -25,6 +25,7 @@
 #include <iomanip>
 #include <algorithm>
 #include <TMath.h>
+#include "TMVA/MethodBDT.h"
 
 // include combinations header (not yet included in boost)
 #include "combination.hpp"
@@ -706,43 +707,7 @@ PFEGammaAlgo(const PFEGammaAlgo::PFEGConfigInfo& cfg) :
   nVtx_(0.0),
   x0inner_(0.0), x0middle_(0.0), x0outer_(0.0),
   excluded_(0.0), Mustache_EtRatio_(0.0), Mustache_Et_out_(0.0)
-{  
-  
-  // Set the tmva reader for electrons
-  tmvaReaderEle_ = new TMVA::Reader("!Color:Silent");
-  tmvaReaderEle_->AddVariable("lnPt_gsf",&lnPt_gsf);
-  tmvaReaderEle_->AddVariable("Eta_gsf",&Eta_gsf);
-  tmvaReaderEle_->AddVariable("dPtOverPt_gsf",&dPtOverPt_gsf);
-  tmvaReaderEle_->AddVariable("DPtOverPt_gsf",&DPtOverPt_gsf);
-  //tmvaReaderEle_->AddVariable("nhit_gsf",&nhit_gsf);
-  tmvaReaderEle_->AddVariable("chi2_gsf",&chi2_gsf);
-  //tmvaReaderEle_->AddVariable("DPtOverPt_kf",&DPtOverPt_kf);
-  tmvaReaderEle_->AddVariable("nhit_kf",&nhit_kf);
-  tmvaReaderEle_->AddVariable("chi2_kf",&chi2_kf);
-  tmvaReaderEle_->AddVariable("EtotPinMode",&EtotPinMode);
-  tmvaReaderEle_->AddVariable("EGsfPoutMode",&EGsfPoutMode);
-  tmvaReaderEle_->AddVariable("EtotBremPinPoutMode",&EtotBremPinPoutMode);
-  tmvaReaderEle_->AddVariable("DEtaGsfEcalClust",&DEtaGsfEcalClust);
-  tmvaReaderEle_->AddVariable("SigmaEtaEta",&SigmaEtaEta);
-  tmvaReaderEle_->AddVariable("HOverHE",&HOverHE);
-//   tmvaReaderEle_->AddVariable("HOverPin",&HOverPin);
-  tmvaReaderEle_->AddVariable("lateBrem",&lateBrem);
-  tmvaReaderEle_->AddVariable("firstBrem",&firstBrem);
-  tmvaReaderEle_->BookMVA("BDT",cfg_.mvaWeightFileEleID.c_str());
-  
-  
-  //Book MVA  
-  tmvaReader_ = new TMVA::Reader("!Color:Silent");  
-  tmvaReader_->AddVariable("del_phi",&del_phi);  
-  tmvaReader_->AddVariable("nlayers", &nlayers);  
-  tmvaReader_->AddVariable("chi2",&chi2);  
-  tmvaReader_->AddVariable("EoverPt",&EoverPt);  
-  tmvaReader_->AddVariable("HoverPt",&HoverPt);  
-  tmvaReader_->AddVariable("track_pt", &track_pt);  
-  tmvaReader_->AddVariable("STIP",&STIP);  
-  tmvaReader_->AddVariable("nlost", &nlost);  
-  tmvaReader_->BookMVA("BDT",cfg_.mvaweightfile.c_str());  
-
+{   
   //Material Map
   TFile *XO_File = new TFile(cfg_.X0_Map.c_str(),"READ");
   X0_sum    = (TH2D*)XO_File->Get("TrackerSum");
@@ -752,8 +717,9 @@ PFEGammaAlgo(const PFEGammaAlgo::PFEGConfigInfo& cfg) :
   
 }
 
-void PFEGammaAlgo::RunPFEG(const reco::PFBlockRef&  blockRef,
-			      std::vector<bool>& active) {  
+void PFEGammaAlgo::RunPFEG(const pfEGHelpers::HeavyObjectCache* hoc,
+                           const reco::PFBlockRef&  blockRef,
+                           std::vector<bool>& active) {  
 
   fifthStepKfTrack_.clear();
   convGsfTrack_.clear();
@@ -765,12 +731,14 @@ void PFEGammaAlgo::RunPFEG(const reco::PFBlockRef&  blockRef,
   // ... will be setable via CFG file parameter
   verbosityLevel_ = Chatty;          // Chatty mode.
   
-  buildAndRefineEGObjects(blockRef);
+  buildAndRefineEGObjects(hoc, blockRef);
 }
 
-float PFEGammaAlgo::EvaluateSingleLegMVA(const reco::PFBlockRef& blockref, 
-					const reco::Vertex& primaryvtx, 
-					unsigned int track_index) {  
+float PFEGammaAlgo::
+EvaluateSingleLegMVA(const pfEGHelpers::HeavyObjectCache* hoc,
+                     const reco::PFBlockRef& blockref, 
+                     const reco::Vertex& primaryvtx, 
+                     unsigned int track_index) {  
   const reco::PFBlock& block = *blockref;  
   const edm::OwnVector< reco::PFBlockElement >& elements = block.elements();  
   //use this to store linkdata in the associatedElements function below  
@@ -814,7 +782,11 @@ float PFEGammaAlgo::EvaluateSingleLegMVA(const reco::PFBlockRef& blockref,
   double vtx_phi=rvtx.phi();  
   //delta Phi between conversion vertex and track  
   del_phi=fabs(deltaPhi(vtx_phi, elements[track_index].trackRef()->innerMomentum().Phi()));  
-  mvaValue = tmvaReader_->EvaluateMVA("BDT");  
+  
+  float vars[] = { del_phi, nlayers, chi2, EoverPt,
+                   HoverPt, track_pt, STIP, nlost };
+
+  mvaValue = hoc->gbrSingleLeg_->GetAdaBoostClassifier(vars);
   
   return mvaValue;
 }
@@ -846,7 +818,8 @@ bool PFEGammaAlgo::isAMuon(const reco::PFBlockElement& pfbe) {
   return false;
 }
 
-void PFEGammaAlgo::buildAndRefineEGObjects(const reco::PFBlockRef& block) {
+void PFEGammaAlgo::buildAndRefineEGObjects(const pfEGHelpers::HeavyObjectCache* hoc,
+                                           const reco::PFBlockRef& block) {
   LOGVERB("PFEGammaAlgo") 
     << "Resetting PFEGammaAlgo for new block and running!" << std::endl;
   _splayedblock.clear();
@@ -929,7 +902,7 @@ void PFEGammaAlgo::buildAndRefineEGObjects(const reco::PFBlockRef& block) {
   // and try to link those in...
   for( auto& RO : _refinableObjects ) {    
     // look for conversion legs
-    linkRefinableObjectECALToSingleLegConv(RO);
+    linkRefinableObjectECALToSingleLegConv(hoc,RO);
     dumpCurrentRefinableObjects();
     // look for tracks that complement conversion legs
     linkRefinableObjectConvSecondaryKFsToSecondaryKFs(RO);
@@ -971,7 +944,7 @@ void PFEGammaAlgo::buildAndRefineEGObjects(const reco::PFBlockRef& block) {
   dumpCurrentRefinableObjects();
 
   // fill the PF candidates and then build the refined SC
-  fillPFCandidates(_refinableObjects,outcands_,outcandsextra_);
+  fillPFCandidates(hoc,_refinableObjects,outcands_,outcandsextra_);
 
 }
 
@@ -1839,7 +1812,8 @@ linkRefinableObjectConvSecondaryKFsToSecondaryKFs(ProtoEGObject& RO) {
 }
 
 void PFEGammaAlgo::
-linkRefinableObjectECALToSingleLegConv(ProtoEGObject& RO) { 
+linkRefinableObjectECALToSingleLegConv(const pfEGHelpers::HeavyObjectCache* hoc,
+                                       ProtoEGObject& RO) { 
   IsConversionTrack<reco::PFBlockElementTrack> isConvKf;
   auto KFbegin = _splayedblock[reco::PFBlockElement::TRACK].begin();
   auto KFend = _splayedblock[reco::PFBlockElement::TRACK].end();  
@@ -1861,8 +1835,9 @@ linkRefinableObjectECALToSingleLegConv(ProtoEGObject& RO) {
     }
     // go through non-conv-identified kfs and check MVA to add conversions
     for( auto kf = notconvkf; kf != notmatchedkf; ++kf ) {
-      float mvaval = EvaluateSingleLegMVA(_currentblock, *cfg_.primaryVtx, 
-                               kf->first->index());
+      float mvaval = EvaluateSingleLegMVA(hoc,_currentblock, 
+                                          *cfg_.primaryVtx, 
+                                          kf->first->index());
       if(mvaval > cfg_.mvaConvCut) {
 	const reco::PFBlockElementTrack* elemaskf =
 	  docast(const reco::PFBlockElementTrack*,kf->first);
@@ -1901,7 +1876,8 @@ linkRefinableObjectSecondaryKFsToECAL(ProtoEGObject& RO) {
 }
 
 void PFEGammaAlgo::
-fillPFCandidates(const std::list<PFEGammaAlgo::ProtoEGObject>& ROs,
+fillPFCandidates(const pfEGHelpers::HeavyObjectCache* hoc,
+                 const std::list<PFEGammaAlgo::ProtoEGObject>& ROs,
 		 reco::PFCandidateCollection& egcands,
 		 reco::PFCandidateEGammaExtraCollection& egxs) {
   // reset output collections
@@ -1974,8 +1950,11 @@ fillPFCandidates(const std::list<PFEGammaAlgo::ProtoEGObject>& ROs,
         const auto &mvavalmapped = RO.singleLegConversionMvaMap.find(kf);
         //FIXME: Abuse single mva value to store both provenance and single leg mva score
         //by storing 3.0 + mvaval
-        float mvaval = mvavalmapped!=RO.singleLegConversionMvaMap.end() ? mvavalmapped->second : 3.0 + EvaluateSingleLegMVA(_currentblock, *cfg_.primaryVtx, 
-                                kf->index());
+        float mvaval = ( mvavalmapped != RO.singleLegConversionMvaMap.end() ? 
+                         mvavalmapped->second : 
+                         3.0 + EvaluateSingleLegMVA(hoc,_currentblock,
+                                                    *cfg_.primaryVtx, 
+                                                    kf->index()) );
         
         xtra.addSingleLegConvTrackRefMva(std::make_pair(kf->trackRef(),mvaval));
       }
@@ -2017,7 +1996,7 @@ fillPFCandidates(const std::list<PFEGammaAlgo::ProtoEGObject>& ROs,
       cand.setP4(p4);   
       cand.setPositionAtECALEntrance(kf->positionAtECALEntrance());
     }    
-    const float ele_mva_value = calculate_ele_mva(RO,xtra);
+    const float ele_mva_value = calculate_ele_mva(hoc,RO,xtra);
     fill_extra_info(RO,xtra);
     //std::cout << "PFEG ele_mva: " << ele_mva_value << std::endl;
     xtra.setMVA(ele_mva_value);    
@@ -2028,7 +2007,8 @@ fillPFCandidates(const std::list<PFEGammaAlgo::ProtoEGObject>& ROs,
 }
 
 float PFEGammaAlgo::
-calculate_ele_mva(const PFEGammaAlgo::ProtoEGObject& RO,
+calculate_ele_mva(const pfEGHelpers::HeavyObjectCache* hoc,
+                  const PFEGammaAlgo::ProtoEGObject& RO,
 		  reco::PFCandidateEGammaExtra& xtra) {
   if( !RO.primaryGSFs.size() ) return -2.0f;
   const PFGSFElement* gsfElement = RO.primaryGSFs.front().first;
@@ -2186,7 +2166,11 @@ calculate_ele_mva(const PFEGammaAlgo::ProtoEGObject& RO,
 		<< " firstBrem " << firstBrem << endl;
       */
       
-      return tmvaReaderEle_->EvaluateMVA("BDT");
+      float vars[] = { lnPt_gsf, Eta_gsf, dPtOverPt_gsf, DPtOverPt_gsf, chi2_gsf,
+                       nhit_kf, chi2_kf, EtotPinMode, EGsfPoutMode, EtotBremPinPoutMode,
+                       DEtaGsfEcalClust, SigmaEtaEta, HOverHE, lateBrem, firstBrem };
+
+      return hoc->gbrEle_->GetAdaBoostClassifier(vars);
     }
   }
   return -2.0f;

--- a/RecoParticleFlow/PFProducer/src/PFEGammaHeavyObjectCache.cc
+++ b/RecoParticleFlow/PFProducer/src/PFEGammaHeavyObjectCache.cc
@@ -1,0 +1,49 @@
+#include "RecoParticleFlow/PFProducer/interface/PFEGammaHeavyObjectCache.h"
+#include "FWCore/ParameterSet/interface/FileInPath.h"
+#include "TMVA/MethodBDT.h"
+#include "TMVA/Reader.h"
+
+namespace pfEGHelpers {
+  HeavyObjectCache::HeavyObjectCache(const edm::ParameterSet& conf) {
+    {
+      const edm::FileInPath& wfile = conf.getParameter<edm::FileInPath>("pf_electronID_mvaWeightFile");
+      // Set the tmva reader for electrons
+      TMVA::Reader tmvaReaderEle_("!Color:Silent");
+      tmvaReaderEle_.AddVariable("lnPt_gsf",&lnPt_gsf);
+      tmvaReaderEle_.AddVariable("Eta_gsf",&Eta_gsf);
+      tmvaReaderEle_.AddVariable("dPtOverPt_gsf",&dPtOverPt_gsf);
+      tmvaReaderEle_.AddVariable("DPtOverPt_gsf",&DPtOverPt_gsf);
+      //tmvaReaderEle_.AddVariable("nhit_gsf",&nhit_gsf);
+      tmvaReaderEle_.AddVariable("chi2_gsf",&chi2_gsf);
+      //tmvaReaderEle_.AddVariable("DPtOverPt_kf",&DPtOverPt_kf);
+      tmvaReaderEle_.AddVariable("nhit_kf",&nhit_kf);
+      tmvaReaderEle_.AddVariable("chi2_kf",&chi2_kf);
+      tmvaReaderEle_.AddVariable("EtotPinMode",&EtotPinMode);
+      tmvaReaderEle_.AddVariable("EGsfPoutMode",&EGsfPoutMode);
+      tmvaReaderEle_.AddVariable("EtotBremPinPoutMode",&EtotBremPinPoutMode);
+      tmvaReaderEle_.AddVariable("DEtaGsfEcalClust",&DEtaGsfEcalClust);
+      tmvaReaderEle_.AddVariable("SigmaEtaEta",&SigmaEtaEta);
+      tmvaReaderEle_.AddVariable("HOverHE",&HOverHE);
+      //   tmvaReaderEle_.AddVariable("HOverPin",&HOverPin);
+      tmvaReaderEle_.AddVariable("lateBrem",&lateBrem);
+      tmvaReaderEle_.AddVariable("firstBrem",&firstBrem);
+      std::unique_ptr<TMVA::IMethod> temp( tmvaReaderEle_.BookMVA("BDT", wfile.fullPath().c_str()) );
+      gbrEle_.reset( new GBRForest( dynamic_cast<TMVA::MethodBDT*>( tmvaReaderEle_.FindMVA("BDT") ) ) );
+    }
+    {
+      const edm::FileInPath& wfile = conf.getParameter<edm::FileInPath>("pf_convID_mvaWeightFile");
+      //Book MVA (single leg)
+      TMVA::Reader tmvaReader_("!Color:Silent");  
+      tmvaReader_.AddVariable("del_phi",&del_phi);  
+      tmvaReader_.AddVariable("nlayers", &nlayers);  
+      tmvaReader_.AddVariable("chi2",&chi2);  
+      tmvaReader_.AddVariable("EoverPt",&EoverPt);  
+      tmvaReader_.AddVariable("HoverPt",&HoverPt);  
+      tmvaReader_.AddVariable("track_pt", &track_pt);  
+      tmvaReader_.AddVariable("STIP",&STIP);  
+      tmvaReader_.AddVariable("nlost", &nlost);  
+      std::unique_ptr<TMVA::IMethod> temp( tmvaReader_.BookMVA("BDT", wfile.fullPath().c_str()) );
+      gbrSingleLeg_.reset( new GBRForest( dynamic_cast<TMVA::MethodBDT*>( tmvaReader_.FindMVA("BDT") ) ) );
+    }
+  }
+}

